lib/stages/receiveFlags.cpp

```diff
@@ -139,23 +155,36 @@ void receiveFlags::main_thread() {
 
         // Copy flags into frame
         flags_lock.lock();
-        update = flags.get_update(ts_frame);
-        if (update.second == nullptr) {
-            ERROR("receiveFlags: Flags for frame {:d} with timestamp {:f} are not in memory. "
-                  "updateQueue is empty",
-                  frame_id_in, ts_to_double(ts_frame));
+
+        // Get and unpack the update
+        const auto& [ts_update, update] = flags.get_update(ts_frame);
+
+        if (update == nullptr) {
+            FATAL_ERROR("receiveFlags: Flags for frame {:d} with timestamp {:f} are not in memory. "
+                        "updateQueue is empty",
+                        frame_id_in, ts_to_double(ts_frame));
+            return;
+        }
+
+        const auto& [state_id, flag_list] = *update;
+
+        std::pair<state_id_t, dset_id_t> key = {state_id, frame_out.dataset_id};
+        if (output_dataset_ids.count(key) == 0) {
+            output_dataset_ids[key] = dm.add_dataset(state_id, frame_out.dataset_id);
         }
-        ts_late = update.first - ts_frame;
+        frame_out.dataset_id = output_dataset_ids[key];
+
+        ts_late = ts_update - ts_frame;
         if (ts_late.tv_sec > 0 && ts_late.tv_nsec > 0) {
```

Comment:
this is probably irrelevant, but shouldn't this be `||`?

Reply:
Hmmmm. Actually it doesn't need to be either. `tv_nsec` is guaranteed to be `>= 0` and so the second step can be skipped.